### PR TITLE
Promote watch e2e test to conformance

### PR DESCRIPTION
**Pull request by jennybuckley**
**What this PR does / why we need it**:
Opening a watch is probably the most widely used pattern for interacting of the kubernetes API (all controllers use it) and needs to be tested by conformance, so its functionality can be relied on in any kubernetes environment.

The intention to have a conformance test for watch was laid out by [this document](https://docs.google.com/document/d/1h2S9ff9N-4MKqfayE3A8TqjD_qIwuND_dAhOAJFxYS0/edit)

The test is not flaky according to [testgrid](https://k8s-testgrid.appspot.com/sig-release-master-blocking#gci-gce&width=5&include-filter-by-regex=api-machinery&include-filter-by-regex=Watchers)

**Release note**:
```release-note
New conformance test added for Watch.
```

/area conformance

### CODE REVIEW

**Comment by bgrant0607** [comment]
Thanks!

I looked at the test PR #60331.

Pods are a bit tricky due to deletion grace period. Was pod chosen deliberately for that reason? I'd think an inert object in the Nucleus like a ConfigMap or Event would be more straightforward.

Also, while testing the basic operations is a reasonable start, I'd like to ensure we test "sync and tail" resumption at a specific resourceVersion.

cc @kubernetes/sig-api-machinery-api-reviews @kubernetes/sig-api-machinery-pr-reviews

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-architecture-pr-reviews @kubernetes/sig-architecture-api-reviews

**Comment by bgrant0607** [comment]
@wojtek-t Thanks for volunteering to take a look at the watch test. I'm happy to iterate if we think the current test is a reasonable MVP, but would like to ensure that the exposed etcd-dependent behaviors are adequately tested. (But not the things we tell users not to depend on, such as resourceVersion sequencing.)

File: test/e2e/apimachinery/watch.go [hunk 1]
```diff
@@ -41,7 +41,13 @@ const (
...
 	f := framework.NewDefaultFramework("watch")
 
-	It("should observe add, update, and delete events on pods", func() {
+	/*
+		    Testname: watch-pods-with-label-selector
+		    Description: Ensure that multiple watchers are able to receive all add,
+			update, and delete events on pods that match a label selector and do
+			not receive events for pods which do not match that label selector.
+	*/
+	framework.ConformanceIt("should observe add, update, and delete events on pods", func() {
```
Comment (wojtek-t) [inline test/e2e/apimachinery/watch.go]: I think for the purpose of conformance tests we need a bit more than what is here.

The scenarios that I would additionally see tested are:

1. check if starting the same watch once all changes are already there will result in the same event (for that purpose we need to start the watch from a given resource version)

2. check if then restarting the watch on RV on which the previous finished will continue delivering events from that point in time.

Comment (jennybuckley) [reply test/e2e/apimachinery/watch.go]: Thank you, this is a very helpful suggestion.

Comment (wojtek-t) [reply test/e2e/apimachinery/watch.go]: So the test you added addresses the first point, but not necessary the second one.

I'm fine with the PR itself, but I'm not convinced we should already mark these tests as conformance.
In other words, I'm fine with merging this PR if you:
- change ConformanceIt to IT
- add a TODO describing the second scenario we want to test

File: test/e2e/apimachinery/watch.go [hunk 2]
```diff
@@ -33,34 +33,43 @@ import (
...
 	f := framework.NewDefaultFramework("watch")
 
-	It("should observe add, update, and delete events on pods", func() {
+	/*
+		    Testname: watch-pods-with-multiple-watchers
+		    Description: Ensure that multiple watchers are able to receive all add,
+			update, and delete events on pods that match a label selector and do
+			not receive events for pods which do not match that label selector.
+	*/
+	framework.ConformanceIt("should observe add, update, and delete events on pods", func() {
```
Comment (timothysc) [inline test/e2e/apimachinery/watch.go]: "should observe add, update, and delete watch notifications on pods"

**Review by timothysc — COMMENTED** [review]
I agree with @wojtek-t 's suggestion and I would also update some of the language on the tests to be clear on the behavior as mentioned in review.

**Comment by jennybuckley** [comment]
/retest
unrelated flakes

File: test/e2e/apimachinery/watch.go [hunk 3]
```diff
@@ -33,158 +31,258 @@ import (
...
+			framework.Failf("Expected notfication to refer to a configmap but got: %v", lastEvent)
+		}
+		testWatchRestarted, err := watchConfigMaps(f, lastEventConfigMap.ObjectMeta.ResourceVersion, watchRestartedLabelValue)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Expecting not to observe duplicate notifications for changes already observed by the first watch")
+		expectNoEvent(testWatchRestarted, watch.Added, testConfigMap)
+		expectNoEvent(testWatchRestarted, watch.Modified, testConfigMapFirstUpdate)
+
+		By("modifying the configmap a second time")
```
Comment (wojtek-t) [inline test/e2e/apimachinery/watch.go]: Please move that higher.

What I would like to ensure is that even if something happened in the meantime (i.e. after watch was stopped but before it was restarted), an event for that will still be delivered.

File: test/e2e/apimachinery/watch.go [hunk 4]
```diff
@@ -33,158 +31,258 @@ import (
...
-		expectEvent(watchB, watch.Added, testPodB)
-		expectEvent(watchAB, watch.Added, testPodB)
-		expectNoEvent(watchA, watch.Added, testPodB)
+		expectEvent(watchB, watch.Added, testConfigMapB)
+		expectEvent(watchAB, watch.Added, testConfigMapB)
+		expectNoEvent(watchA, watch.Added, testConfigMapB)
 
-		err = c.CoreV1().Pods(ns).Delete(testPodB.GetName(), nil)
+		By("deleting configmap B and ensuring the correct watchers observe the notification")
+		err = c.CoreV1().ConfigMaps(ns).Delete(testConfigMapB.GetName(), nil)
```
Comment (wojtek-t) [inline test/e2e/apimachinery/watch.go]: One more thing that I would like to see tested (sorry that I didn't mention it on the first time, but I realized in the meantime that it's an extremely important case and we used to have bugs about that in the past is):

- you create an object with label - add watch event should be send
- you delete that label (or change its value) - delete watch event should be send
- you add the label back - add event should be send this time
- you delete the object - delete watch event should be send.

This should probably be a separate (fourth) test.

**Review by wojtek-t — COMMENTED** [review]
Thanks - it looks much better now.
I added one comment and a request for one more test (it should be fairly simple to add it with all helpers that you already have).

**Comment by AishSundar** [comment]
@jennybuckley I had a chance to talk to 1.11 RT about this PR. Considering we will be entering CodeFreeze next Tuesday, 6/5, we should ideally aim for this PR to merge by EOD Thursday 5/31 for this to make into 1.11. We will then have time for a couple of test runs before code freeze and branch cut. thanks.

File: test/e2e/apimachinery/watch.go (hunk 3 above)
Comment (jennybuckley) [reply test/e2e/apimachinery/watch.go]: Done, for some reason github moved your comment to a different test

File: test/e2e/apimachinery/watch.go (hunk 4 above)
Comment (jennybuckley) [reply test/e2e/apimachinery/watch.go]: also done

**Comment by jennybuckley** [comment]
@AishSundar that should be enough time, I think this PR completely done now

**Comment by timothysc** [comment]
Please squash your commits.

**Comment by jennybuckley** [comment]
@timothysc squashed

**Comment by wojtek-t** [comment]
thanks!

/lgtm
/approve

**Comment by AishSundar** [comment]
@deads2k to take a look

**Comment by bgrant0607** [comment]
cc @smarterclayton

**Comment by bgrant0607** [comment]
In the future, please write e2e tests first, and create a PR later to add the test to conformance, once it proves stable. If this test proves to be flaky, it will need to be removed from the conformance suite.

That approach also decouples the production of useful tests from their promotion to conformance.

**Comment by AishSundar** [comment]
@bgrant0607 Ack on splitting the addition of e2e tests and promotion to Conformance as 2 separate PRs and we will be enforcing that as a process going forward. Since this PR was in works for a few months now, it seemed to have followed the older process. I will keep an eye out for how these tests fare in Conformance suite before deciding if it should go in 1.11 or should be removed.

**Comment by AishSundar** [comment]
The new tests seem to be passing for a few runs now, both on [master](https://k8s-testgrid.appspot.com/conformance-all#GCE,%20master%20(dev)&include-filter-by-regex=Watchers) and [1.11 branches.](https://k8s-testgrid.appspot.com/conformance-all#GCE,%20v1.11%20(dev)&include-filter-by-regex=Watchers)

**Pull request merged**